src/xmerl_c14n.erl

```diff
@@ -0,0 +1,527 @@
+    Attrs;
+add_defaultns(Attrs, _DefaultNs, NewDefaultNs) ->
+    [#xmlAttribute{ name = xmlns
+                  , pos = 1
+                  , value = atom_to_list(NewDefaultNs)
+                  , normalized = false
+                  } | Attrs].
+
+get_defaultns(#xmlElement{ nsinfo = {Prefix, _Name}
+                         , namespace = Nss
+                         }) ->
+    {Prefix, URI} = lists:keyfind(Prefix, 1, Nss#xmlNamespace.nodes),
+    URI.
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-spec build_nsinfo(#xmlNamespace{}, #xmlElement{}) -> #xmlElement{}.
+build_nsinfo(Ns, Attr = #xmlAttribute{name = Name}) ->
+    case string:tokens(atom_to_list(Name), ":") of
+        [NsPrefix, Rest] -> Attr#xmlAttribute{namespace = Ns, nsinfo = {NsPrefix, Rest}};
+        _ -> Attr#xmlAttribute{namespace = Ns}
+    end;
+build_nsinfo(Ns, Elem = #xmlElement{name = Name, content = Kids, attributes = Attrs}) ->
+    Elem2 = case string:tokens(atom_to_list(Name), ":") of
+        [NsPrefix, Rest] -> Elem#xmlElement{namespace = Ns, nsinfo = {NsPrefix, Rest}};
+        _ -> Elem#xmlElement{namespace = Ns}
+    end,
+    Elem2#xmlElement{attributes = [build_nsinfo(Ns, Attr) || Attr <- Attrs],
+                    content = [build_nsinfo(Ns, Kid) || Kid <- Kids]};
+build_nsinfo(_Ns, Other) -> Other.
+
+canon_name_test() ->
+    "urn:foo:Blah" = canon_name("foo", "Blah", #xmlNamespace{nodes = [{"foo", 'urn:foo:'}]}),
+    "urn:foo:Blah" = canon_name("foo", 'Blah', #xmlNamespace{nodes = [{"foo", 'urn:foo:'}]}),
+    {'EXIT', {{ns_not_found, _, _}, _}} = (catch canon_name("foo", "Blah", #xmlNamespace{})),
+    "urn:bar:Blah" = canon_name("bar", "Blah", #xmlNamespace{nodes = [{"bar", 'urn:bar:'}]}).
+canon_name_attr_test() ->
+    "urn:foo:Blah" = canon_name(#xmlAttribute{name = 'Blah', nsinfo = {"foo", "Blah"}, namespace = #xmlNamespace{nodes = [{"foo", 'urn:foo:'}]}}).
+canon_name_elem_test() ->
+    "urn:foo:Blah" = canon_name(#xmlElement{name = 'Blah', nsinfo = {"foo", "Blah"}, namespace = #xmlNamespace{nodes = [{"foo", 'urn:foo:'}]}}).
+
+canon_name_default_ns_test() ->
+    {Doc, []} = xmerl_scan:string("<foo:a xmlns:foo=\"urn:foo:\"><b xmlns=\"urn:bar:\"></b></foo:a>", [{namespace_conformant, true}]),
+    A = #xmlElement{content = [B]} = Doc,
+    "urn:foo:a" = canon_name(A),
+    "urn:bar:b" = canon_name(B).
+
+needed_ns_test() ->
+    Ns = #xmlNamespace{nodes = [{"foo", 'urn:foo:'}, {"bar", 'urn:bar:'}]},
+
+    E1 = build_nsinfo(Ns, #xmlElement{name = 'foo:Blah', attributes = [#xmlAttribute{name = 'bar:name', value="foo"}]}),
+    ["bar", "foo"] = lists:sort(needed_ns(E1, [])),
+
+    E2 = build_nsinfo(Ns, #xmlElement{name = 'Blah', attributes = [#xmlAttribute{name = 'bar:name', value = "foo"}]}),
+    ["bar"] = needed_ns(E2, []),
+
+    E3 = build_nsinfo(Ns, #xmlElement{name = 'Blah', attributes = [#xmlAttribute{name = 'name', value = "foo"}], content = [#xmlElement{name = 'foo:InnerBlah'}]}),
+    [] = needed_ns(E3, []),
```

Comment:
Why is foo not needed? If `needed_ns/2` only returns the namespaces used in the root level element it was given, it should clearly state that. Reading this test and the doc for `needed_ns` I would have expected "foo" to be required.